### PR TITLE
Shadowlings get their own lighting overlay

### PR DESCRIPTION
**Pull request by monster860**
http://imgur.com/7Usv3M3

Active when weakeyes = 1. Based off of the AI's /datum/camerachunk code.

:cl:
rscadd: Shadowlings have an overbright overlay on nondark turfs
/:cl:

### CODE REVIEW

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+				t.shadowling_bright_overlay = image('icons/effects/shadowling_bright.dmi', t.bright_overlay_container, "overlay", 15)
+				t.shadowling_bright_overlay.blend_mode = BLEND_ADD
+				t.shadowling_bright_overlay.mouse_opacity = 0
+				if(t.lighting_overlay)
+					t.lighting_overlay.update_shadowling_overlay(t)
+
+			overlays += t.shadowling_bright_overlay
+
+/mob/proc/update_shadowling_chunks()
+	if(!weakeyes)
```
Comment (Fox-McCloud) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: not everything that has weakeyes is a shadowling....

**Comment by Fox-McCloud** [comment]
I don't like this, at all.

**Comment by monster860** [comment]
You cant see in dark and stay out of light at the same time though.

**Comment by Regen1** [comment]
While it makes much more sense being able to tell where the light is, I think you should make the changelog a lot more clear.

They're supposed to be short and descriptive of what you changed without having to see it for yourself
"Shadowlings can now tell more easily where the light is" sounds cryptic and not really tell the reader anything about what exactly has been changed.

**Comment by monster860** [comment]
Is that a better changelog?

**Comment by ZomgPonies** [comment]
I like this a lot. It makes shadowlings/shadowpeople actually playable now with no more excuse for complaining

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	if(visible > 0)
+		visible--
+
+/turf/var/image/shadowling_bright_overlay
+/turf/var/atom/movable/bright_overlay_container
+
+/datum/shadowchunk/New(loc, x, y, z)
+
+	// 0xf = 15
+	x &= ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: Don't use hex. It's harder to read, and all of our code uses directly defined numbers.

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	y &= ~0xf
+
+	src.x = x
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: `src.` is unnecessary

Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: I take that back. It isn't unnecessary here. Change the argument names.

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+			overlays += t.shadowling_bright_overlay
+
+/mob/proc/update_shadowling_chunks()
+	if(!weakeyes)
+		for(var/datum/shadowchunk/C in visible_shadow_chunks)
+			C.remove(src)
+		return
+	// 0xf = 15
+	var/x1 = max(0, x - 16) & ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: No hex, use real numbers.

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	var/y1 = max(0, y - 16) & ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: :eyes:

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	var/x2 = min(world.maxx, x + 16) & ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: :eyes:

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	var/y2 = min(world.maxy, y + 16) & ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: :eyes:

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		var/datum/shadowchunk/C = chunk
+		C.add(src)
+
+/mob/Move()
+	. = ..()
+	update_shadowling_chunks()
+
+var/list/shadow_chunks = list()
+proc/get_shadow_chunk(x, y, z)
+	x &= ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: :eyes::eyes:

File: code/game/gamemodes/shadowling/shadowling_sight.dm [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	y &= ~0xf
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: :eyes:

File: code/modules/lighting/lighting_overlay.dm [hunk 10]
```diff
@@ -60,6 +60,14 @@
 		needs_update = 1
 		lighting_update_overlays += src
 
+/atom/movable/lighting_overlay/proc/update_shadowling_overlay(var/turf/T)
+	if(T.shadowling_bright_overlay)
```
Comment (tigercat2000) [inline code/modules/lighting/lighting_overlay.dm]: This won't work if we ever update the lighting system. Won't support making it harder to modify.

File: code/game/gamemodes/shadowling/shadowling_sight.dm (hunk 2 above)
Comment (monster860) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: I copypasted most of this from camerachunk.

**Comment by monster860** [comment]
A lot of this is copypaste from the AI's /datum/camerachunk, including the hex numbers.

File: code/modules/lighting/lighting_overlay.dm (hunk 10 above)
Comment (monster860) [inline code/modules/lighting/lighting_overlay.dm]: How would it make updatinf lighting system harder?

**Comment by tigercat2000** [comment]
Just because you copy-pasted it doesn't mean you shouldn't fix it.

File: code/modules/lighting/lighting_overlay.dm (hunk 10 above)
Comment (tigercat2000) [inline code/modules/lighting/lighting_overlay.dm]: If/when PJB releases his goon-lights based system, lighting doesn't work the same, based on a fairly different concept. This code won't work then, and I don't want to have anything to worry about maintaining with it. Entirely my own subjective thing with this, the code will work with the current system.

Comment (monster860) [inline code/modules/lighting/lighting_overlay.dm]: This could be modified to work with goonlights.

**Comment by DaveTheHeadcrab** [comment]
Anyone?
I actually really like this feature.

File: code/game/gamemodes/shadowling/shadowling_sight.dm (hunk 1 above)
Comment (KasparoVy) [inline code/game/gamemodes/shadowling/shadowling_sight.dm]: Maybe it's better this way. *shrug*

**Comment by Regen1** [comment]
@Fox-McCloud @monster860 @TheDZD 

Any progress?

**Comment by tigercat2000** [comment]
@Regen1 this won't be merged as-is, it's a way overly complicated system which applies to all weak-eyes antagonists 

it needs to either be a simple night-vision spell for shadowlings only or not at all

**Comment by Fox-McCloud** [comment]
Closeing for reasons started by @tigercat2000, refusal to update the changes, and stagnation.

**Pull request closed without merging**